core/primitives/src/receipt.rs

```diff
@@ -210,55 +210,25 @@ impl BorshDeserialize for Receipt {
     /// Deserialize based on the first and second bytes of the stream. For V0, we do backward compatible deserialization by deserializing
     /// the entire stream into V0. For V1, we consume the first byte and then deserialize the rest.
     fn deserialize_reader<R: Read>(reader: &mut R) -> std::io::Result<Self> {
-        let u1 = u8::deserialize_reader(reader)?;
-        let u2 = u8::deserialize_reader(reader)?;
-        let u3 = u8::deserialize_reader(reader)?;
-        let u4 = u8::deserialize_reader(reader)?;
         // This is a ridiculous hackery: because the first field in `ReceiptV0` is an `AccountId`
         // and an account id is at most 64 bytes, for all valid `ReceiptV0` the second byte must be 0
         // because of the little endian encoding of the length of the account id.
         // On the other hand, for `ReceiptV1`, since the first byte is 1 and an account id must have nonzero
         // length, so the second byte must not be zero. Therefore, we can distinguish between the two versions
         // by looking at the second byte.
 
-        let read_predecessor_id = |buf: [u8; 4], reader: &mut R| -> std::io::Result<AccountId> {
-            let str_len = u32::from_le_bytes(buf);
-            let mut str_vec = Vec::with_capacity(str_len as usize);
-            for _ in 0..str_len {
-                str_vec.push(u8::deserialize_reader(reader)?);
-            }
-            AccountId::try_from(String::from_utf8(str_vec).map_err(|_| {
-                Error::new(ErrorKind::InvalidData, "Failed to parse AccountId from bytes")
-            })?)
-            .map_err(|e| Error::new(ErrorKind::InvalidData, e.to_string()))
-        };
+        let u1 = u8::deserialize_reader(reader)?;
+        let u2 = u8::deserialize_reader(reader)?;
+
+        let prefix = if u2 == 0 { vec![u1, u2] } else { vec![u2] };
```

Comment:
mini nit: I slightly dislike that this check (u2 == 0) happens in two places. Maybe consider moving it inside the respective if cases? Or make it into a variable `let is_v0 = u2 == 0`?

Reply:
nice catch. 2nd way is better